src/string_lib.c

```diff
@@ -0,0 +1,48 @@
+#include <stdio.h>
+
+size_t str_len(const char *s)
+{
+  if (NULL == s)
+  {
+    return 0;
+  }
+
+  size_t length = 0;
+  while (*s != 0)
+  {
+    length++;
+    s++;
+  }
+
+  return length;
+}
+
+int str_cmp(const char *s1, const char *s2)
+{
+  int res = 0;
+
+  while (*s1 != 0 && *s2 != 0)
+  {
+    s1++, s2++;
+  }
+
+  if (*s1 != 0 && *s2 == 0)
+  {
+    res++;
+    while (*(++s1) != 0)
+    {
+      res++;
+    }
+  }
+
+  if (*s1 == 0 && *s2 != 0)
+  {
+    res--;
+    while (*(++s2) != 0)
+    {
+      res--;
+    }
+  }
```

Comment:
This is not necessary.
Once you have done the first loop there is no need to loop anymore.
You just need to compare `*s1` and `*s2`.

```suggestion
  if (*s1 == *s2) {
    return 0;
  }
  if (*s1 < *s2) {
    return -1;
  }
  return 1; // No need for a comparison here as it is the last possible case!
```

Reply:
"less than, equal to, or greater than zero" confuses me a lot.